File: Python/Problems/Problems/problem696.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter

from .problem import Problem

class Problem696(Problem):
    def test(self) -> bool:
        s = "00110011"

        expected = 6

        result = self.countBinarySubstrings(s)

        print(result)

        return result == expected

    def countBinarySubstrings(self, s: str) -> int:
        n = len(s)
        last = 0
        while last < n - 1 and s[last] == s[last + 1]:
            last += 1
        result = 0
        prev = last + 1
        for i in range(last + 1, n - 1):
            if s[i] != s[i + 1]:
                curr = i - last
                result += min(curr, prev)
                last = i
                prev = curr
        result += min(prev, n - last - 1)

        return result

File: C++/Problems/Problems/problem_696.cpp
#include "problem_696.h"

bool problem_696::test() {
    std::string s = "00110011";

    int expected = 6;

    int result = countBinarySubstrings(s);

    std::cout << result << std::endl;

    return result == expected;
}

int problem_696::countBinarySubstrings(std::string s) {
    int n = s.size();
    int last = 0;
    while (last < n - 1 && s[last] == s[last + 1])
        last++;
    int result = 0;
    int prev = last + 1;
    for (int i = last + 1; i < n - 1; ++i)
    {
        if (s[i] != s[i + 1])
        {
            int curr = i - last;
            result += std::min(curr, prev);
            last = i;
            prev = curr;
        }
    }
    result += std::min(prev, n - last - 1);

    return result;
}
